### ISSUE TITLE
Cases where more than one "Meaningful Sequence" is required for total meaning

### ISSUE THREAD

**Issue by brennanyoung** [w3c/wcag#1062]
I'd like to raise a problem I have been having making accessible flowcharts which I have traced (partly, but significantly) to the WCAG documentation.

The document for “Understanding Success Criterion 1.3.2: Meaningful Sequence” says

> For clarity:
> 
> 1.	Providing a particular linear order is only required where it affects meaning.
> 2.	There may be more than one order that is "correct" (according to the WCAG 2.0 definition).
> 3.	Only one correct order needs to be provided.

A flowchart, **by definition**, is likely to have more than one meaningful sequence, except in the most basic examples.

Furthermore, the representation of **multiple** meaningful sequences is **the prime use case** for flowcharts. If only one meaningful sequence is represented, the meaning of the flowchart is compromised or lost.

Therefore, **if "only one correct order" is provided, meaning is affected**. Points of clarity 1 and 3 are in contradiction in such cases.

Consider a life-saving algorithm for heart failure. (A real world case from my own work). The choice box might say "Is the victim breathing? YES/NO". If the (currently) WCAG compliant version of the flowchart only offers the "YES" branch, how will the disabled user discover what to do in the "NO" case? This loss of meaning is so fundamental that the entire flowchart fails to communicate.

Note: ARIA has no viable specification for representing branching, browsable content. A mechanism may be hacked together using operable roles, but for the most part (beyond choosing branches) flowcharts are read-only, browse-mode content. I have had my 'accessible' flowcharts rejected at WCAG audits because I have used ARIA features that pertain to operable roles. But what alternatives are available today for meeting SC 1.3.2 than to use the tab sequence to 'browse' a flowchart, with some kind of operable role (e.g. a listbox) for the branches?

Additionally, the boxes may very well contain rich text, which calls for the AT to be in browse mode. In our case, the flowchart boxes contain lists, headings and other rich text. How will I make these semantics available to disabled users?

There's a need for further clarity in WCAG, work in ARIA and implementation in UAs and AT to resolve this problem area properly. It is (of course) a subset of the larger unsolved problem of browsable data visualisations.

I would appreciate if WCAG could offer some guidance in this matter. At the very least, it should acknowledge the shortcomings of the current specifications, so that authors and auditors have more reasonable expectations about what may be achieved in this area.

**Comment by mraccess77** [comment]
Given the SC allows for one correct order to pass then I'd imagine anyone failing you on reading order is not understanding the SC correctly.    If you didn't provide a way to transverse the chart with keyboard or other mechanism then I'd likely say it was a 2.4.3 focus order issue.  If you do provide a way to navigate the chart with the keyboard then you'd likely pass that.  As you mentioned you'd also have to meet SC 4.1.2 with name, role, state, etc.   In terms of allowing users to read text in steps -- there is a document role that can be used which allows screen readers to enter browse mode.  You could also have a keystroke that might pop up a modal which could put the user in browse mode to read the text and then close the modal to return to focus mode for navigation in the chart.   I do think this type of thing can be made accessible with screen readers and also be conformant.  I do agree that aria-flowto was not designed for complex situations and rearranging the whole accessibility tree will never be a good solution -- so a solution like I mention going in and out of focus mode is best.   You'd want to describe the in and outlets of any digraph and provide summary information about the chart as well.

**Comment by brennanyoung** [comment]
I think the original issue was flagged under 2.4.3 because the auditors simply didn't try pressing TAB. When I pointed out how it worked, they said we needed to provide instructions because users wouldn't expect the flowchart to be 'tabbable', and would expect a text alternative instead. They then said it was a Name, Role, Value violation because of the operable/browsable conflict. (At that time I was using list and listitem for the boxes which were unavoidably placed in the tab sequence). I didn't find this advice very helpful.

I am aware of the document role, and I have in fact decided to use it in this case, although I have been advised **not** to use it multiple times by a11y experts. Reasons are rarely given, but usually it is something like "it's too complicated". Hmm.

But to stay focused on SC 1.3.2 (which is my main point here) - the "understanding" doc contains points of 'clarity' which come into contradiction when representing flowcharts. I am calling for this contradiction to be addressed in such cases.

**Comment by bruce-usab** [comment]
@brennanyoung, do you have some suggested edits for us to consider?

> Therefore, if "only one correct order" is provided, meaning is affected. Points of clarity 1 and 3 are in contradiction in such cases.

I don't agree that there is a contradiction.  It is a minimum requirement.  Worse case is that by providing **a** single reading order it implies that there is **only one** reading order.  But this can happen with lots of kinds of content.

**Comment by JAWS-test** [comment]
> ARIA has no viable specification for representing branching, browsable content

I think [aria-flowto](https://www.w3.org/TR/wai-aria-1.2/#aria-flowto) is theoretically perfect for this. Unfortunately, the AT support is sometimes not so good (see for JAWS: https://github.com/FreedomScientific/VFO-standards-support/issues/326)

**Comment by brennanyoung** [comment]
Indeed. `aria-flowto` is obviously the Right Thing here, which is why I am so dismayed at the talk of deprecation. And yes, support is poor. JAWS is the only AT that handles it, as far as I know.

**Comment by aarongustafson** [comment]
@brennanyoung Have you considered filing bugs against other screen readers to encourage support for `aria-flowto`?

**Comment by brennanyoung** [comment]
I added a comment to the existing NVDA issue about `aria-flowto` but yes, I should report the problem to Apple and Microsoft too. We also still need to establish a keyboard idiom for how to choose amongst multiple ids listed in the attribute value.
https://github.com/nvaccess/nvda/issues/4049

**Comment by patrickhlauke** [comment]
i still don't get how `aria-flowto` would help you here, as it indicates a single set order. even if that was supported, your idea seems to rely on a lot of far more fundamental magic to allow for dynamic choice of different paths which would then require the flow to dynamically change, or am i missing something?

**Comment by brennanyoung** [comment]
Yes, offering a way for authors and users to change the browse flow is exactly what I am calling for. This is also what is promised by the latest `aria-flowto` spec today. Really, I just want the AT vendors to follow the spec. Apart from JAWS, they don't even try.

The [spec ](https://www.w3.org/TR/wai-aria/#aria-flowto) for `aria-flowto` is quite clear that multiple (space separated) ids may be included in the attribute value:

> In the case of one or more IDREFS, user agents or assistive technologies SHOULD give the user the option of navigating to any of the targeted elements. The name of the path can be determined by the name of the target element of the aria-flowto attribute.

This implies a branching set order, rather than a single set order, no? I mean, there is a single flowchart, but the browsing 'pathway' is dynamic. Changing the browse flow is exactly what happens when an AT navigates via element type (e.g. next heading, next table, or whatever), so this should not be an exotic idea.

What's missing is for the ATs to offer some special announcements and keyboard shortcuts for choosing amongst those branches.

**Here's a suggested pattern (criticisms welcome!):**
- User browses to an element with `aria-flowto="x y z"`. A tab index is not required.
- AT announces (e.g.) "flows to 1 of 3:" followed by the accessible name of element x. 
- With a given input (e.g. the **+** key), the AT announces something like "flow to 2 of 3" followed by the accessible name of element y. 
- The + key (or whatever it may be) cycles through the choices, announcing the accessible name of each.
- Shift+ could go the other way. 'Wrapping around' to be decided, perhaps by AT vendors.
- Pressing the 'browse to next element' key shortcut (e.g. down arrow on NVDA) should then browse to whichever of the flowto ids is 'pointed at', exactly as if it appeared next in the DOM.

Alternative patterns might be imagined, but this suggestion is, I believe, consistent with the current spec.

Tricky question: Should there be some mechanism so that the currently indicated 'destination' element knows that it is targeted before we browse to it? Should it receive some kind of event or message? How might we indicate it visually?

In many cases (such as footnotes), the first id in the list should probably (for UX reasons) point to the next element in the DOM, and the subsequent item(s) should point to any alternative 'destination(s)', so that the 'default' flow is the one in the DOM. The jump to the footnote would thus require one additional keystroke, but 'plain old browsing' to the next chunk of content would continue to be handled by the same input mechanisms as it does today.

**Comment by brennanyoung** [comment]
I'd like to add that for branching flow diagrams, this needs to work with **SVG** as well as (or perhaps more so) than HTML, and therefore it might be appropriate to include it in the [ARIA graphics spec](https://github.com/w3c/graphics-aria/issues/1)

**Comment by patrickhlauke** [comment]
ah, my mistake, i was going from memory from an early definition of `aria-flowto` that only allowed a single idref. this request now makes more sense.

**Comment by brennanyoung** [comment]
Question: If we have more than one 'flow targets' from our current location, and one of the targets is the next element in the DOM. (The default 'next' piece of content). Should we require that this next adjacent element is included in the `aria-flowto` attribute?
Or should there always be a default: proceed according to DOM browse order, so that 'continue' is always available? 

The spec suggests (but not too clearly) that even with only a single IDREF value indicated `aria-flowto`, a choice is offered: Continue browsing from here, or jump somewhere else.

**Comment by mraccess77** [comment]
One potential issue with flow to is that it seems to only be forward not backward   Folks will want to go back and without flow to referring to prior direction content and having clear naming it is less optimal.    In certain directed graphs paths can be in out or in and out.     It would be necessary to allow traversal to all but mark them as such.

**Comment by brennanyoung** [comment]
@mraccess77 your point is worth exploring.

In my own (Tab-sequence dependent) prototypes, I implemented a 'history' stack (similar to the browser's Back button), so that shift-Tab handled the backwards direction. Theoretically, the AT could offer something similar.

Footnotes are the prime use case for 'getting back' after a browse flow detour. A footnote might be referenced from more than one place, so a single ID `aria-flowfrom` ([as offered in prefixed form by Microsoft](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/x-ms-aria-flowfrom)) is probably not enough.

But if we can imagine a choosing mechanism for flowto, we can perhaps imagine the same mechanism to get back again?

**Comment by brennanyoung** [comment]
Boiling this issue down, I arrive at these three facts:

- SC 1.3.1 requires that meaning is preserved, regardless of presentation
- SC 1.3.2 requires only one meaningful sequence
- If meaning is lost _because_ only one meaningful sequence is offered by a given presentation, this **necessarily** violates SC 1.3.1

The section in 1.3.2 headed with the words "For clarity" is misleading. It states "Only one correct order needs to be provided." but it does not point out that there are cases where this may violate SC 1.3.1 - that's not clarity, it's an important omission.

I would very much like to see this resolved by pointing out this possible violation of SC 1.3.1 in the 'intent' copy for SC 1.3.2, in much the same way that the 'intent' copy for SC 2.4.6 devotes three paragraphs to other success criteria which 'fill the gaps' not covered by SC 2.4.6

**Comment by bruce-usab** [comment]
@brennanyoung do you have an example of your tab-able flow chart posted somewhere?  Authors not doing an adequate job with flowcharts is pretty common.  Having something that was interactive instead of static could be a great alternative!  (I am so sorry that I didn't ask for this back in March!)

> If meaning is lost because only one meaningful sequence is offered by a given presentation, this necessarily violates SC 1.3.1

Sorry to say, I don't agree with your conclusion in this case.

I have seen examples where content provides offer a static PDF as the conforming alternative version for an interactive tutorial.  When then challenged as why not provide _only_ the static PDF **to everyone** they quickly articulate why that is not acceptable -- which, of course, lets me explain why the static PDF is not good enough for PWD.

IMHO, if you want to make the case that multiple meaningful sequences is essential to the presentation, then you need more than a semantic argument. Static descriptions of static images have a long tradition of being sufficient.
